apollo-server/v3/integration-testing/server.test.ts

```diff
@@ -0,0 +1,68 @@
+// @ts-check
+// we import our typeDefs and resolvers for creating an Apollo Server instance for testing
+const { typeDefs, resolvers, ApolloServer } = require('./server.ts');
+const express = require('express');
+const { createServer } = require('http');
+
+// we will use supertest to test our server
+const request = require('supertest');
+
+// this is the query we use for our test
+const queryData = {
+  query: `query sayHello($name: String) {
+    hello(name: $name)
+  }`,
+  variables: { name: 'world' },
+};
+
+// create a new Apollo server instance for testing
+const createTestApolloServer = async () => {
+  const testServer = new ApolloServer({
+    typeDefs,
+    resolvers,
+  });
+
+  await testServer.start();
+  return testServer;
+};
+
+// create a new http server for testing (we pass in our test instance of Apollo Server)
+const createTestHttpServer = async (server) => {
+  // create our test
+  const app = express();
+
+  await server.applyMiddleware({
+    app,
+    path: '/graphql',
+  });
+
+  const httpServer = createServer(app);
+  const port = process.env.PORT || 3000;
```

Comment:
It's probably better to model running tests against servers that pick their own ephemeral port rather than using a fixed port.  For this, you can do `listen({ port: 0 })` and then (after waiting for the `listening` event to fire) you can ask the server for its address.  eg, you can see how the standalone server `apollo-server` does it at https://github.com/apollographql/apollo-server/blob/444c403011209023b5d3b5162b8fb81991046b23/packages/apollo-server/src/index.ts#L125-L131 and https://github.com/apollographql/apollo-server/blob/444c403011209023b5d3b5162b8fb81991046b23/packages/apollo-server/src/index.ts#L59-L83

Alternatively, you might just be happier using `apollo-server` (the standalone server) in this example instead of `apollo-server-express`! This means you won't have to set up an express app and HTTP server, and it has a nice `listen` function that's async and returns useful things. On the other hand, I guess you can't use that with `supertest` which expects an `http.Server` which is hidden from you here. Ah well, I guess you do have to do it yourself. (Or use an HTTP test library other than supertest which takes an URL.)

... ok, digging more deeply into `supertest`, looks like there's two options which would both be reasonable:

(a) You actually can pass an URL to `supertest`, not just a function or http.Server. I get this from the source, not the docs, sadly. https://github.com/visionmedia/supertest/blob/c2fd9ece02b365fecbf36ae2a4f7c0bf528486ff/lib/test.js#L32-L33 That means you can use the standalone `apollo-server`, call `const { url } = await server.listen({ port: 0 })`, and pass that `url` to `supertest`

(b) If you pass a non-listening http.Server to supertest, it will do the dance for you of "listen on an ephemeral port, connect to that port, close the server later".  So you can use `apollo-server-express` as you're doing here and just skip the `httpServer.listen` call, and pass the `httpServer` to `request()`.

Oh, and what you're doing below with `request(expressApp)` is definitely *not* what you want to do: supertest will interpret that as "you haven't made an http server at all, start a new one and listen on it", so now you have two http servers for no reason.

(a) is going to lead to the least code in the test suite, and will be best for folks just getting started who are still on the standalone server. (b) might be helpful for the common case that folks have graduated to apollo-server-express.  I'd lean towards (a) for now? Could always make two examples later, one for each package. (In AS4 both of these will be in the same `@apollo/server` package and will have different exported names instead of being two things called `ApolloServer` in different packages.)

Reply:
Fantastic! I agree A will provide the clearest example, thank you **so much** for digging into this and providing such thorough feedback. This example is so much shorter and easier to read now!  🤩